src/Data/Array/Mutable/Linear.hs

```diff
@@ -120,7 +120,8 @@ read = Unsafe.toLinear readUnsafe
     readUnsafe :: Array a -> Int -> (Array a, Unrestricted a)
     readUnsafe arr@(Array size mutArr) ix
       | indexInRange size ix =
-          (arr, Unrestricted $ Unsafe.readMutArr mutArr ix)
+          let val = Unsafe.readMutArr mutArr ix
+          in  val `seq` (arr, Unrestricted val)
```

Comment:
I'd do:
```suggestion
          let !val = Unsafe.readMutArr mutArr ix
          in (arr, Unrestricted val)
```
I find it a bit clearer.

Interestingly, this is stricter than it needs to be. That is, if the `arr` stores `undefined` at `ix`, this version of `read arr ix` will loop, whereas it should be `(arr, undefined)`. At least if our arrays are to have the same semantics as the mutable arrays of base.

Here is an idea, however: instead of storing `a`, let's store `(# a #)` instead. (either we do this directly in `Unsafe.Array`, or we define `Array a` to be an `Unsafe.Array (# a #)`. I don't know. Then we would have
```suggestion
          let (# val #)  = Unsafe.readMutArr mutArr ix
          in  (arr, Unrestricted val)
```

As a bonus, we can't get this wrong: the types force us to force the value.

Reply:
Alas, this array implementation looks awfully easy to get wrong.

Reply:
> Alas, this array implementation looks awfully easy to get wrong.

What is “this” in your sentence?

Reply:
I hadn't imagined that making an array implementation that used linear types would be so tightly concerned with the evaluation order. Putting arrays and linear types together doesn't produce an obviously correct implementation. In retrospective, it is to be expected, of course.

"This" didn't referred to your fix proposal, but rather to the whole endeavor of providing linear arrays.

Reply:
> Interestingly, this is stricter than it needs to be.
> Here is an idea, however: instead of storing a, let's store (# a #) instead.

I agree that this is stricter. But, in this case what we need to force is `Unsafe.readMutArr mutArr ix` call rather than `ix` itself, so actually I was thinking about the opposite, maybe we should store a `Identity a` inside instead. So then we could store bottoms inside. However it does introduce an overhead, so I'm not sure if we should.

Reply:
Why `Identity a`? It has the exact same strictness as `a`. Maybe you meant a `data` version of `Identity`, to add an indirection. In which case, yes, there is a cost. That's why I suggested `(# a #)`: exactly the same representation as `a`, but always forced to constructor form `(# x #)`, without `x` being forced.

In fact, now that I think of it, we may not even need to _store_ a `(# a #)`, if `Unsafe.readMutArr` makes sure to return a `(# a #)` (one has to place the constructor in the right location though).

Reply:
I thought forcing an unboxed tuple would also force the value, if that's not the case, that'd be good.

Reply:
An unboxed tuple is _always_ forced. But it can contain a thunk. This is, indeed, the point.